### PR TITLE
Fixup meson build

### PR DESCRIPTION
**Pull request by tristan957**
The meson build had gotten a little out of hand. It needed to be cleaned
up and have its errors fixed. This should enable lz4 to switch to Meson
at any time should the need ever arise.

### CODE REVIEW

**Comment by tristan957** [comment]
@Cyan4973 Here is what I got regarding fixing Meson. The only thing that I didn't try to port were the tests. I compiled the executables, but the tests are so entrenched in the Makefile that I just lost motivation. I see we don't even run the tests on the Meson build, so I don't think it will be an issue.

CC @eli-schwartz

**Comment by tristan957** [comment]
Wondering if I should do something like this: https://github.com/hse-project/hse/blob/master/subprojects/packagefiles/lz4/lib/meson.build#L5. That is my custom lz4 build.

File: contrib/meson/meson.build
Comment (tristan957) [inline contrib/meson/meson.build]: @Cyan4973 could you pick out the spdx identifiers from this list? https://spdx.org/licenses/ that I can put here?

File: contrib/meson/meson/programs/meson.build
Comment (tristan957) [inline contrib/meson/meson/programs/meson.build]: @eli-schwartz could you review this usage of install_symlink?

**Comment by tristan957** [comment]
Wondering if we could forego an lz4 release and just create a new release in wrapdb with the new Meson build. That would solve my issues.

**Comment by tristan957** [comment]
I also left the original copyright headers in place although the code is vastly different. Do you want me to add myself to the contributor headers?

**Comment by Cyan4973** [comment]
> I also left the original copyright headers in place although the code is vastly different. Do you want me to add myself to the contributor headers?

sounds fair

File: contrib/meson/meson.build
Comment (Cyan4973) [reply contrib/meson/meson.build]: The library is licensed as BSD, so it would be `BSD-2-Clause-Patent`,
but the CLI (Command Line Interface) is licensed as GPLv2, which would be `GPL-2.0-or-later`.

**Comment by tristan957** [comment]
> Wondering if I should do something like this: https://github.com/hse-project/hse/blob/master/subprojects/packagefiles/lz4/lib/meson.build#L5. That is my custom lz4 build.

I don't think is necessary. I ran a test to see if an executable would link without lz4 APIs going through the PLT with a static lz4 library, and it seemed good to go.

**Comment by tristan957** [comment]
Anyways I feel really good about this PR and I am happy to mark it off my list of things to do.

**Comment by tristan957** [comment]
https://github.com/mesonbuild/wrapdb/pull/179#issuecomment-942543236

hmmm...

File: contrib/meson/meson/contrib/gen_manual/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/contrib/gen_manual/meson.build:33]: The use of `@INPUT@` and the input kwarg is a definite improvement.

File: contrib/meson/meson/examples/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/examples/meson.build]: This seems to be optimized for a case that won't come up. I think you can make this a lot simpler.

```meson

examples = {
  'printVersion': 'printVersion.c',
}
```
And later on...

File: contrib/meson/meson/examples/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/examples/meson.build]: ... you can do this:


```meson

foreach e, src : examples
  executable(
    e,
    lz4_source_root / 'examples' / src,
```

File: contrib/meson/meson/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/meson.build]: You cannot use fs.parent, because the fs module is not available until meson 0.53, but before Meson 0.61 we forgot to add a FeatureNew warning about it.

However it is not strictly needed. Instead of every file using `../../../` with different levels, you can define it once, here, but using `../` instead of `fs.parent()`

`meson.current_source_dir() / '../../../'`

File: contrib/meson/meson/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/meson.build:27]: You dropped `-pedantic` and it's not implied by the current project default value for the `warning_level` option. I see you consolidated the individual flags from each target individually, though.

File: contrib/meson/meson/lib/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/lib/meson.build]: You set this elsewhere using add_project_arguments.

File: contrib/meson/meson/programs/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/programs/meson.build]: Why do you need to extract all objects here? Guessing because the build never used gnu symbol visibility before, *and* the executable is relying on symbols that aren't exported with LZ4LIB_API?

File: contrib/meson/meson/programs/meson.build
Comment (eli-schwartz) [reply contrib/meson/meson/programs/meson.build]: `pointing_to` needs to be a string, not a meson `executable()` object.

So, since you want to create `lz4c` as a program which is a symlink to the `lz4` program, just use `pointing_to: 'lz4'`. (Meson then creates a relative symlink that tries to find the lz4 destination in the same directory as the symlink itself.)

Comment (eli-schwartz) [reply contrib/meson/meson/programs/meson.build]: I'd use a foreach loop. `foreach alias: ['lz4c', 'lz4cat', 'unlz4']`, create two symlinks, one in the bindir pointing to `'lz4'`, one in the mandir named `alias + '.1'` pointing to `'lz4.1'`

File: contrib/meson/meson/programs/meson.build
Comment (eli-schwartz) [inline contrib/meson/meson/programs/meson.build]: A meson `install_dir` never needs to be joined with get_option('prefix'), since internally it already uses the equivalent of join_paths or the `/` operator  to join any relative value with the prefix.

You only need to join these together if you need the absolute path of e.g. bindir as a string, for example to print in a message or to add to a configuration_data() and write into a configured header file.

File: contrib/meson/meson/programs/meson.build
Comment (tristan957) [reply contrib/meson/meson/programs/meson.build]: Exactly. @Cyan4973 any thoughts on this?

File: contrib/meson/meson/lib/meson.build
Comment (tristan957) [reply contrib/meson/meson/lib/meson.build]: `compile_args` is exported in `declare_dependency(compile_args:)`. Not used for c_args. Unless you had a different intention with your comment?

File: contrib/meson/meson/programs/meson.build
Comment (eli-schwartz) [reply contrib/meson/meson/programs/meson.build]: It seems a bit unusual to only create the aliases when the user has a new enough version of Meson. Though there is no denying it's a lot cleaner to use the Meson function.

File: contrib/meson/meson/lib/meson.build
Comment (eli-schwartz) [reply contrib/meson/meson/lib/meson.build]: No, I just mistook what its purpose was here. Sorry!

...

Does that mean this needs to be in the generated pkgconfig file too?

Comment (tristan957) [reply contrib/meson/meson/lib/meson.build]: @Cyan4973 would be the authority here. Maybe the behavior I wrote is not what should be exhibited.

File: contrib/meson/meson/programs/meson.build
Comment (tristan957) [reply contrib/meson/meson/programs/meson.build]: Since this is a community contributed port and the symlinks are deprecated as far as I can tell, I just wanted to say screw it and use the Meson functions.

Comment (eli-schwartz) [reply contrib/meson/meson/programs/meson.build]: Yeah, it's not super important for them to exist I guess so having it be less well supported is probably fine.

File: contrib/meson/meson/lib/meson.build
Comment (Cyan4973) [reply contrib/meson/meson/lib/meson.build]: Note that I'm not familiar with `meson`.
Consequently, I'm unaware of the consequences of employing `compile_args`.
From a `Makefile` perspective, something like `-DLZ4_MEMORY_USAGE=#` is supposed to belong to the `CPPFLAGS` variable, rather than the `CFLAGS` one. The difference can be substantial since `CFLAGS` are ignored at link stage, but `CPPFLAGS` are not. It's unclear if this level of differentiation exists at `meson` level.

Then, if, as mentioned by @eli-schwartz, the flags are then also present in `pkgconfig`, well this is unusual, but it seems it would allow the library to be compiled with a different hash table size than the default size (directly defined in the source code). This is pretty advanced stuff, that is rarely ever used. Furthermore, in most cases, this size is abstracted away and does not matter. So it takes a very advanced scenario for this case to matter. And such advanced scenario may more likely happen in the embedded world, where a `pkgconfig` is less likely to be a thing.
So, to summarize, yes, in some rare cases, it could prove useful. But as importantly, in all cases I can think of, and presuming that the `pkgconfig` remains in sync with the installed library, it should not be detrimental.

File: contrib/meson/meson/programs/meson.build
Comment (Cyan4973) [reply contrib/meson/meson/programs/meson.build]: I don't remember the details, but I know that the CLI executable is designed to be linked to the _static_ variant of the library, meaning it is not restricted to "public" symbols. One such example is the `xxhash` library, which is not exposed on the public side of `liblz4`, but is employed internally to calculate checksums. The same function is used either as a random number generator or as a checksum for different purposes within the CLI.

Comment (eli-schwartz) [reply contrib/meson/meson/programs/meson.build]: The most elegant solution here, IMHO, is to do a bit of project restucturing until you end up with:
- `libcommon.a`: a static library containing internal symbols shared between the library and the CLI
- `liblz4.so`/`liblz4.a`: a shared/static library containing public symbols of the library only. It links to libcommon
- `lz4`: a CLI executable that links to both libcommon and to liblz4

So, the CLI executable would be free to link to the shared variant of the library, and it ***is*** restricted to public symbols, but any time you decide you need a private symbol, you just move the private symbol to libcommon. This also has a neat side effect: you get to see which functions are broken out into this internal library, and which presumably represent weak points in the `liblz4.so` API that don't fully meet the needs of your own CLI. Who knows -- those might turn out to be things that merit polishing and stabilization.

The CLI executable then prefers to link to public symbols from liblz4, which depending on the vendor's choice may be static or shared, because some vendors have decided for some reason they like shared libraries ;)

But, for any non-public symbols, the CLI executable just falls back to statically linking them from the convenience library `libcommon`.

You get the best of both worlds. The downside is it requires more work.

File: contrib/meson/meson/lib/meson.build
Comment (tristan957) [reply contrib/meson/meson/lib/meson.build]: `compile_args` tells a project using this lz4 dep to compile with the listed `compile_args` in the compiler invocation. Based on what you said this isn't expected nor popular. Therefore I will remove it!

File: contrib/meson/meson/programs/meson.build
Comment (tristan957) [reply contrib/meson/meson/programs/meson.build]: I want to copy essentially what upstream is currently doing. So for me I see two options other than the re-org suggested by @eli-schwartz. Link against `liblz4_internal_dep` which is an internal static library or continue to do what I am doing currently which is just extracting objects from the static library.

File: contrib/meson/meson/meson.build
Comment (tristan957) [reply contrib/meson/meson/meson.build]: I get the following error:

```
../meson/lib/meson.build:39:0: ERROR: Tried to form an absolute path to a source dir.
You should not do that but use relative paths instead.
```

File: contrib/meson/meson/meson.build
Comment (tristan957) [reply contrib/meson/meson/meson.build:27]: Covered by moving to warning_level=3

File: contrib/meson/meson/meson.build
Comment (tristan957) [reply contrib/meson/meson/meson.build]: For now I just bumped to minimum Meson version to 0.53 as mentioned.

Comment (eli-schwartz) [reply contrib/meson/meson/meson.build]: (I didn't test it, sorry.)

Oh right... it should be an absolute path in both cases.

The difference is that, given rootdir: `/tmp/lz4/contrib/meson`:
- in the `fs.parent(rootdir)` case it is: `/tmp/lz4/lib`
- in the `currentdir / '../../../'` case it is: `/tmp/lz4/contrib/meson/meson/../../../lib`

When integrating lz4 into the wrapdb, we move the meson.build into the top level of the lz4 source tree so `/tmp/lz4` will be the rootdir and both methods will spawn an error.

Sorry, but you cannot simplify this. Please continue to hardcode a set number of `../../../` in the header of each file. It is necessary as long as the Meson files are in `contrib/meson/` instead of in `.` and there is no getting around it (well, other than putting them back where Meson expects them to be).

**Comment by tristan957** [comment]
Alright hopefully that is the last push.

**Review by eli-schwartz — APPROVED** [review]
I think this answers all the feedback I had.

**Comment by tristan957** [comment]
@Cyan4973 anything else you need from me on this PR?

**Comment by Cyan4973** [comment]
It would be great if the `meson` CI test was successful (green) after this PR
https://github.com/lz4/lz4/runs/5128636505?check_suite_focus=true

**Comment by tristan957** [comment]
Whoops. My apologies. Will look into it.

**Comment by tristan957** [comment]
I just pushed a fix and have raised an internal issue with Meson, but that shouldn't affect lz4 in any way

**Comment by tristan957** [comment]
Cool the CI is now all good to go as far as this PR is concerned

**Comment by tristan957** [comment]
@vtorri isn't `pic` controlled by `b_staticpic` from Meson? Don't see anything that needs changes on this side of things. Don't see why that would need to be the default, but I am also not familiar with `pic` to begin with.

**Comment by tristan957** [comment]
@Cyan4973 any more feedback for this PR? I got @eli-schwartz to approve since he is a maintainer of Meson and it's wrapdb, so he would be fairly authoritative on this matter.

**Review by Cyan4973 — APPROVED** [review]
This looks good to me

**Comment by Cyan4973** [comment]
Thanks for the `meson` script's improvement @tristan957 !

**Comment by tristan957** [comment]
You are welcome. I'll definitely be giving this a go as soon as we get a new release on wrapdb. I may come back with more feedback, but I think this is a much better base than what we had before.

**Comment by tristan957** [comment]
So @eli-schwartz informed me that I can't see the results of my labor easily unless there is a 1.9.4 release of lz4. Would this be possible? Otherwise I can wait until the next release that fits within your schedule.

**Comment by Cyan4973** [comment]
I don't have any availability to work on a new release of `lz4` before Q2, at best.

**Comment by tristan957** [comment]
That's ok. No big deal. Look forward to the next release

**Comment by tristan957** [comment]
Hey @Cyan4973. Wondering if there was anything on your roadmap for a release including this work any time soon. I know Q2 isn't over just yet though

**Comment by Cyan4973** [comment]
There is no release planned _soon_.
I'll need a few days of focus to get to it.
Maybe during the coming Summer.

**Pull request merged**
